src/adapters/net.cpp

```diff
@@ -324,68 +350,131 @@ namespace net {
-      return;
-    }
-    // Check if the driver supplies the quality level value
-    if (stats.qual.updated & IW_QUAL_LEVEL_INVALID) {
-      return;
-    }
+    auto status = nla_get_u32(bss[NL80211_BSS_STATUS]);
 
-    // Check if the link quality has been uodated
-    if (stats.qual.updated & IW_QUAL_QUAL_UPDATED) {
-      m_linkquality.val = stats.qual.qual;
-      m_linkquality.max = range.max_qual.qual;
-    }
+    return status == NL80211_BSS_STATUS_ASSOCIATED || status == NL80211_BSS_STATUS_IBSS_JOINED;
+  }
 
-    // Check if the signal strength has been uodated
-    if (stats.qual.updated & IW_QUAL_LEVEL_UPDATED) {
-      m_signalstrength.val = stats.qual.level;
-      m_signalstrength.max = range.max_qual.level;
+  /**
+   * Set the ESSID
+   */
+  void wireless_network::parse_essid(struct nlattr** bss) {
+    if (bss[NL80211_BSS_INFORMATION_ELEMENTS] == nullptr) {
+      m_essid.clear();
+    } else {
+      // Information Element ID from ieee80211.h
+      #define WLAN_EID_SSID 0
 
-      // Check if the values are defined in dBm
-      if (stats.qual.level > range.max_qual.level) {
-        m_signalstrength.val -= 0x100;
-        m_signalstrength.max = (stats.qual.level - range.max_qual.level) - 0x100;
+      auto ies = static_cast<uint8_t*>(nla_data(bss[NL80211_BSS_INFORMATION_ELEMENTS]));
+      auto ies_len = nla_len(bss[NL80211_BSS_INFORMATION_ELEMENTS]);
+      const auto hdr_len = 2;
+
+      while (ies_len > hdr_len && ies[0] != WLAN_EID_SSID) {
+        ies_len -= ies[1] + hdr_len;
+        ies += ies[1] + hdr_len;
+      }
+
+      if (ies_len < hdr_len || ies_len < ies[1] + hdr_len) {
+        m_essid.clear();
+      } else {
+        m_essid = string(reinterpret_cast<char*>(ies + hdr_len));
       }
     }
   }
 
-  // }}}
-}
+  /**
+   * Set frequency
+   */
+  void wireless_network::parse_frequency(struct nlattr** bss) {
+    if (bss[NL80211_BSS_FREQUENCY] != nullptr) {
+      // in MHz
+      m_frequency = static_cast<int>(nla_get_u32(bss[NL80211_BSS_FREQUENCY]));
+    }
+  }
+
+  /**
+   * Set device driver quality values
+   */
+  void wireless_network::parse_quality(struct nlattr** bss) {
+    if (bss[NL80211_BSS_SIGNAL_UNSPEC] != nullptr) {
+      // Signal strength in unspecified units, scaled to 0..100 (u8)
+      m_linkquality.val = nla_get_u8(bss[NL80211_BSS_SIGNAL_UNSPEC]);
+      m_linkquality.max = 100;
+    }
+  }
+
+  /**
+   * Set the signalstrength
+   */
+  void wireless_network::parse_signal(struct nlattr** bss) {
+    if (bss[NL80211_BSS_SIGNAL_MBM] != nullptr) {
+      // signalstrength in dBm
+      int signalstrength = static_cast<int>(nla_get_u32(bss[NL80211_BSS_SIGNAL_MBM])) / 100;
+
+      // WiFi-hardware usually operates in the range -100dBm to -20dBm.
+      // For a more intuitive result, clamp the result to a
+      // signal range, where 100% are reachable without directly
+      // sitting on top of the AP
+      const int hardware_max = -40;
```

Comment:
IMO if we modify the values too much it might confuse users if they compare it to other sources.
From `nl80211_xbm_to_percent` it seems like -20 and -90 dBm are the commonly used values, so it might be best to use those instead.

Reply:
This just came up in #1834. Though [`nl80211_xbm_to_percent`](https://github.com/NetworkManager/NetworkManager/blob/master/src/platform/wifi/nm-wifi-utils-nl80211.c#L372) in the NetworkManager code says it goes from `-20` to `-90`, the way the percentage is calculated, it actually goes from `-120dBm` to `-20dBm`. `-90dBm` corresponds to 30%.

But I think the -20 to -90 range is good.